File: cpp/examples/ode_secir_parameter_study.cpp
/*
* Copyright (C) 2020-2025 MEmilio
*
* Authors: Daniel Abele, Martin J. Kuehn
*
* Contact: Martin J. Kuehn <Martin.Kuehn@DLR.de>
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
#include "memilio/compartments/simulation.h"
#include "memilio/config.h"
#include "ode_secir/model.h"
#include "ode_secir/parameters_io.h"
#include "ode_secir/parameter_space.h"
#include "memilio/compartments/parameter_studies.h"
#include "memilio/mobility/metapopulation_mobility_instant.h"
#include "memilio/io/result_io.h"

/**
 * @brief creates xml file with a single run parameter study with std 0 (used to save parameters of individual runs)
 * @param filename Name of file
 * @param params Secir parameters used during run
 * @param t0 starting point of simulation
 * @param tmax end point of simulation
 */
mio::IOResult<void> write_single_run_result(const size_t run, const mio::osecir::Simulation<ScalarType>& sim)
{
    std::string abs_path;
    BOOST_OUTCOME_TRY(auto&& created, mio::create_directory("results", abs_path));

    if (run == 0) {
        std::cout << "Results are stored in " << abs_path << '\n';
        if (!created) {
            std::cout << "Directory already exists, files from previous runs will be overwritten." << '\n';
        }
    }

    //write sampled parameters for this run
    auto node_filename = mio::path_join(abs_path, "Parameters_run" + std::to_string(run) + ".json");
    BOOST_OUTCOME_TRY(mio::write_json(node_filename, sim.get_result()));

    //write results for this run
    std::vector<mio::TimeSeries<ScalarType>> all_results;
    std::vector<int> ids;

    BOOST_OUTCOME_TRY(mio::save_result({sim.get_result()}, {0}, (int)sim.get_model().parameters.get_num_groups().get(),
                                       mio::path_join(abs_path, ("Results_run" + std::to_string(run) + ".h5"))));

    return mio::success();
}

int main()
{
    mio::set_log_level(mio::LogLevel::debug);

    ScalarType t0   = 0;
    ScalarType tmax = 50;
    ScalarType dt   = 0.1;

    ScalarType cont_freq = 10; // see Polymod study

    ScalarType num_total_t0 = 10000, num_exp_t0 = 100, num_inf_t0 = 50, num_car_t0 = 50, num_hosp_t0 = 20,
               num_icu_t0 = 10, num_rec_t0 = 10, num_dead_t0 = 0;

    mio::osecir::Model<ScalarType> model(1);
    mio::AgeGroup num_groups = model.parameters.get_num_groups();
    ScalarType fact          = 1.0 / (ScalarType)(size_t)num_groups;

    auto& params = model.parameters;

    params.set<mio::osecir::ICUCapacity<ScalarType>>(std::numeric_limits<ScalarType>::max());
    params.set<mio::osecir::StartDay<ScalarType>>(0);
    params.set<mio::osecir::Seasonality<ScalarType>>(0);

    for (auto i = mio::AgeGroup(0); i < num_groups; i++) {
        params.get<mio::osecir::TimeExposed<ScalarType>>()[i]            = 3.2;
        params.get<mio::osecir::TimeInfectedNoSymptoms<ScalarType>>()[i] = 2.;
        params.get<mio::osecir::TimeInfectedSymptoms<ScalarType>>()[i]   = 6.;
        params.get<mio::osecir::TimeInfectedSevere<ScalarType>>()[i]     = 12;
        params.get<mio::osecir::TimeInfectedCritical<ScalarType>>()[i]   = 8;

        model.populations[{i, mio::osecir::InfectionState::Exposed}]            = fact * num_exp_t0;
        model.populations[{i, mio::osecir::InfectionState::InfectedNoSymptoms}] = fact * num_car_t0;
        model.populations[{i, mio::osecir::InfectionState::InfectedSymptoms}]   = fact * num_inf_t0;
        model.populations[{i, mio::osecir::InfectionState::InfectedSevere}]     = fact * num_hosp_t0;
        model.populations[{i, mio::osecir::InfectionState::InfectedCritical}]   = fact * num_icu_t0;
        model.populations[{i, mio::osecir::InfectionState::Recovered}]          = fact * num_rec_t0;
        model.populations[{i, mio::osecir::InfectionState::Dead}]               = fact * num_dead_t0;
        model.populations.set_difference_from_group_total<mio::AgeGroup>({i, mio::osecir::InfectionState::Susceptible},
                                                                         fact * num_total_t0);

        params.get<mio::osecir::TransmissionProbabilityOnContact<ScalarType>>()[i] = 0.05;
        params.get<mio::osecir::RelativeTransmissionNoSymptoms<ScalarType>>()[i]   = 0.67;
        params.get<mio::osecir::RecoveredPerInfectedNoSymptoms<ScalarType>>()[i]   = 0.09;
        params.get<mio::osecir::RiskOfInfectionFromSymptomatic<ScalarType>>()[i]   = 0.25;
        params.get<mio::osecir::SeverePerInfectedSymptoms<ScalarType>>()[i]        = 0.2;
        params.get<mio::osecir::CriticalPerSevere<ScalarType>>()[i]                = 0.25;
        params.get<mio::osecir::DeathsPerCritical<ScalarType>>()[i]                = 0.3;
    }

    params.apply_constraints();

    mio::ContactMatrixGroup<ScalarType>& contact_matrix = params.get<mio::osecir::ContactPatterns<ScalarType>>();
    contact_matrix[0]                                   = mio::ContactMatrix<ScalarType>(
        Eigen::MatrixX<ScalarType>::Constant((size_t)num_groups, (size_t)num_groups, fact * cont_freq));

    mio::osecir::set_params_distributions_normal<ScalarType>(model, t0, tmax, 0.2);

    auto write_parameters_status = mio::write_json("parameters.json", model);
    if (!write_parameters_status) {
        std::cout << "Error writing parameters: " << write_parameters_status.error().formatted_message();
        return -1;
    }

    //create study
    auto num_runs = size_t(1);
    // mio::ParameterStudy2<mio::osecir::Simulation<ScalarType>, mio::osecir::Model<ScalarType>, ScalarType>
    //     parameter_study(model, t0, tmax, dt, num_runs);

    auto parameter_study =
        mio::make_parameter_study<mio::osecir::Simulation<ScalarType>>(model, t0, tmax, dt, num_runs);

    //run study
    auto sample_graph = [](const auto& model_, ScalarType t0_, ScalarType dt_, size_t) {
        mio::osecir::Model<ScalarType> copy = model_;
        mio::osecir::draw_sample(copy);
        return mio::osecir::Simulation<ScalarType>(std::move(copy), t0_, dt_);
    };
    auto handle_result = [](auto&& sim, auto&& run) {
        auto write_result_status = write_single_run_result(run, sim);
        if (!write_result_status) {
            std::cout << "Error writing result: " << write_result_status.error().formatted_message();
        }
        return 0; //Result handler must return something, but only meaningful when using MPI.
    };
    parameter_study.run(sample_graph, handle_result);

    return 0;
}

File: pycode/examples/simulation/ode_secir_parameter_study.py
#############################################################################
# Copyright (C) 2020-2025 MEmilio
#
# Authors:
#
# Contact: Martin J. Kuehn <Martin.Kuehn@DLR.de>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#############################################################################

import argparse

import numpy as np

import memilio.simulation as mio
import memilio.simulation.osecir as osecir


def run_ode_secir_parameter_study():
    """ """
    # setup basic parameters
    num_groups = 6
    model = osecir.Model(num_groups)

    for i in range(num_groups):
        group = mio.AgeGroup(i)

        model.parameters.TimeExposed[group] = 3.2
        model.parameters.TimeInfectedNoSymptoms[group] = 2.
        model.parameters.TimeInfectedSymptoms[group] = 6
        model.parameters.TimeInfectedSevere[group] = 12
        model.parameters.TimeInfectedCritical[group] = 8

        model.populations[group, osecir.InfectionState.Exposed] = 100
        model.populations[group, osecir.InfectionState.InfectedNoSymptoms] = 50
        model.populations[group,
                          osecir.InfectionState.InfectedNoSymptomsConfirmed] = 0
        model.populations[group, osecir.InfectionState.InfectedSymptoms] = 50
        model.populations[group,
                          osecir.InfectionState.InfectedSymptomsConfirmed] = 0
        model.populations[group, osecir.InfectionState.InfectedSevere] = 20
        model.populations[group, osecir.InfectionState.InfectedCritical] = 10
        model.populations[group, osecir.InfectionState.Recovered] = 10
        model.populations[group, osecir.InfectionState.Dead] = 0
        model.populations.set_difference_from_group_total_AgeGroup(
            (group, osecir.InfectionState.Susceptible), 10000)

        model.parameters.TransmissionProbabilityOnContact[group].set_distribution(
            mio.ParameterDistributionUniform(0.1, 0.2))
        model.parameters.RelativeTransmissionNoSymptoms[group] = 0.67
        model.parameters.RecoveredPerInfectedNoSymptoms[group] = 0.09
        model.parameters.RiskOfInfectionFromSymptomatic[group] = 0.25
        model.parameters.SeverePerInfectedSymptoms[group] = 0.2
        model.parameters.CriticalPerSevere[group] = 0.25
        model.parameters.DeathsPerCritical[group] = 0.3

    model.parameters.ContactPatterns.cont_freq_mat = mio.ContactMatrixGroup(
        4, num_groups)
    model.parameters.ContactPatterns.cont_freq_mat[0] = mio.ContactMatrix(
        np.ones((num_groups, num_groups))*0.5)
    model.parameters.ContactPatterns.cont_freq_mat[1] = mio.ContactMatrix(
        np.ones((num_groups, num_groups))*0.5)
    model.parameters.ContactPatterns.cont_freq_mat[2] = mio.ContactMatrix(
        np.ones((num_groups, num_groups))*0.5)
    model.parameters.ContactPatterns.cont_freq_mat[3] = mio.ContactMatrix(
        np.ones((num_groups, num_groups))*0.5)
    model.parameters.ContactPatterns.cont_freq_mat.add_damping(
        mio.Damping(np.ones((num_groups, num_groups))*0.7, 7.0))
    print(model.parameters.ContactPatterns.cont_freq_mat[1].baseline)
    # process the result of one run

    def handle_result(graph, run_idx):
        """

        :param graph: 
        :param run_idx: 

        """
        group = mio.AgeGroup(0)
        print("run {} with infection rate {:.2G}".format(handle_result.c, graph.get_node(
            0).property.model.parameters.TransmissionProbabilityOnContact[group].value))
        print("compartments at t = {}:".format(
            graph.get_node(0).property.result.get_time(0)))
        print(graph.get_node(0).property.result.get_value(0))
        print("compartments at t = {}:".format(
            graph.get_node(0).property.result.get_last_time()))
        print(graph.get_node(0).property.result.get_last_value())
        handle_result.c += 1
    handle_result.c = 0

    # study the effect of different infection rates

    model.apply_constraints()

    graph = osecir.ModelGraph()
    graph.add_node(0, model)
    graph.add_node(1, model)
    mobility_coefficients = 0.01 * np.ones(model.populations.numel())
    for i in range(num_groups):
        flat_index = model.populations.get_flat_index(
            osecir.MultiIndex_PopulationsArray(mio.AgeGroup(i), osecir.InfectionState.Dead))
        mobility_coefficients[flat_index] = 0
    graph.add_edge(0, 1, mobility_coefficients)
    graph.add_edge(1, 0, mobility_coefficients)

    study = osecir.GraphParameterStudy(
        graph, t0=0, tmax=10, dt=0.5, num_runs=3)
    study.run(handle_result)


if __name__ == "__main__":
    arg_parser = argparse.ArgumentParser(
        'ode_secir_parameter_study',
        description='Example demonstrating ensemble runs of a ODE SECIHURD model.')
    args = arg_parser.parse_args()
    run_ode_secir_parameter_study()
